### DISCUSSION TITLE
[PSD12] Exercise 3.22

### DISCUSSION THREAD

**Discussion by matypist** [sapienzastudentsnetwork/psd2223#183] [Esercizi (A-L)]
**Exercise 3.22** Complete a state transition table and output table for the FSM in **Figure 3.69**. Write Boolean equations for the next state and output and sketch a schematic of the FSM.

![image](https://user-images.githubusercontent.com/69389461/210066349-d1995ba2-0b40-432a-b76b-cd6e25c9165f.png)

- _**N.B.**: L'esercizio originale del libro prevederebbe un inizio di consegna diverso:_

	**Exercise 3.22** Describe in words what the state machine in Figure 3.69 does. Using binary state encodings, complete [...]

**Comment by Elia-Belli** [comment]
![Immagine 2023-01-17 122132](https://user-images.githubusercontent.com/107944476/212886353-eb111c88-e7c7-4691-a407-09d6307c0e34.png)
$S_0' = \bar S_1 \bar S_0 A$
$S_1' = S_0 B$
$Q = S_1$
![C](https://user-images.githubusercontent.com/107944476/212887920-e177110d-47a4-4486-9ea6-9e2c8de08daa.jpg)

**Comment by FeddyLix17** [comment]
@Elia-Belli non capisco una cosa, le equazioni di **S'<sub>1</sub>** e di **Q** non dovrebbero essere rispettivamente S&#x0304;<sub>1</sub>S<sub>0</sub>B e S<sub>1</sub>S&#x0304;<sub>0</sub> ?

**Comment by fraxb04** [comment]
@FedVlogger17 
@Elia-Belli 
seguendo questo ragionamento mi viene questo circuito:
ps: y = Q, ed ho dimenticato di indicare il CLK, ed il reset 
:)
![image](https://user-images.githubusercontent.com/122573871/214283634-641de5fb-56b3-4412-98f6-c07d4a7c7ab4.png)

**Comment by Elia-Belli** [comment]
> @Elia-Belli non capisco una cosa, le equazioni di **S'1** e di **Q** non dovrebbero essere rispettivamente S̄1S0B e S1S̄0 ?

Sono presenti stati irragiungibili e una codifica per l'output non è utilizzata, quindi li ho considerati come don't care e semplificato le funzioni.

Versione di @fraxb04 corretta, ma non minima

**Comment by fraxb04** [comment]
@Elia-Belli in che senso stati irraggiungibili, e quale codifica dell'output non è utilizzata??

**Comment by Elia-Belli** [comment]
> @Elia-Belli in che senso stati irraggiungibili, e quale codifica dell'output non è utilizzata??

Per codificare gli stati usiamo 2 bit, quindi ipoteticamente ne possiamo realizzare 4, in questo caso ne usiamo 3 e quello che corrisponderebbe a $S3=11$ è irraggiungibile; mentre l'output inutilizzato è quello che corrisponde appunto a $11$.

**Comment by fraxb04** [comment]
@Elia-Belli 
infatti io non ho inserito nessun S3 neanche codificato come 11

**Comment by Elia-Belli** [comment]
> @Elia-Belli infatti io non ho inserito nessun S3 neanche codificato come 11

appunto, allora puoi usarlo come Don't Care, fai una K-map per ogni funzione e ti risulterà evidente.

**Comment by fraxb04** [comment]
@Elia-Belli 
ok capito, quindi in generale occorre farlo per tutte le volte che avanza qualche stato?
grazie mille
EDIT: ho provato a fare la mappa K ma non capisco come considerare i don't care agli input...... per caso puoi allegare una foto della tua K-map?

**Comment by Elia-Belli** [comment]
> @Elia-Belli ok capito, quindi in generale occorre farlo per tutte le volte che avanza qualche stato? grazie mille EDIT: ho provato a fare la mappa K ma non capisco come considerare i don't care agli input...... per caso puoi allegare una foto della tua K-map?

Se vuoi ottenere delle funzioni minime usare le k-map è il modo migliore, anche se in questo esercizio non era richiesto.
Allego la K-map di $S1'$, le variabili sono, in ordine, $S1$, $S0$, $A$, $B$:
![Immagine 2023-01-24 151725](https://user-images.githubusercontent.com/107944476/214319060-e2e3d156-30e4-41da-8ef0-3e52d81289b5.png)

**Comment by fraxb04** [comment]
@Elia-Belli 
ma come hai trasformato 1X e 0X in 00 (intendo AB)??

**Comment by Elia-Belli** [comment]
> @Elia-Belli ma come hai trasformato 1X e 0X in 00 (intendo AB)??

Beh, se hai un minterm $000X$ che vale $1$, allora indipendetemente da $B$ quello varrà sempre $1$, ovvero sia il minterm $0000$ che $0001$ valgono $1$.
